src/charm.py

```diff
@@ -610,14 +697,13 @@ def _test_database_connectivity(self):
             A tuple of connectivity as bool and error message as str, error message will be
             an empty string if charm can connect to the database.
         """
-        db_info = self._current_effective_db_info()
         try:
             # TODO: add database charset check later
             cnx = mysql.connector.connect(
-                host=db_info["DB_HOST"],
-                database=db_info["DB_NAME"],
-                user=db_info["DB_USER"],
-                password=db_info["DB_PASSWORD"],
+                host=self._current_effective_db_info.hostname,
```

Comment:
What if `self._current_effective_db_info` is `None`?

Reply:
I'm actually surprised `mypy` didn't complain, something to check?

Reply:
This is something that I want to refactor, currently there cannot exist a case in which the `self._current_effective_db_info` can be None since it cannot progress after the `if not self._current_effective_db_info` check in `def _core_reconciliation(self)` method.